{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The history saving thread hit an unexpected error (DatabaseError('database disk image is malformed')).History will not be written to the database.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import collections\n",
    "sys.path.append('/cellar/users/aklie/opt/gene_program_evaluation')\n",
    "sys.path.append('/cellar/users/aklie/opt/gene_program_evaluation/app/')\n",
    "import mudata\n",
    "import scanpy as sc\n",
    "\n",
    "from utils import count, count_unique\n",
    "from utils import load_config\n",
    "\n",
    "from parse import (\n",
    "    parse_loadings,\n",
    "    parse_obs_memberships,\n",
    "    parse_software_versions\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pipeline outputs\n",
    "path_config = \"/cellar/users/aklie/opt/gene_program_evaluation/examples/report/iPSC_EC/factor_analysis/factor_analysis_100/report.yaml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = load_config(path_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse config for paths\n",
    "path_evaluation_outs = config[\"path_evaluation_outs\"]\n",
    "path_mdata = config[\"path_mdata\"]\n",
    "path_evaluation_config = config[\"path_evaluation_config\"]\n",
    "path_report_out = config[\"path_report_out\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse config for other parameters\n",
    "data_key = config[\"data_key\"]\n",
    "prog_keys = config[\"prog_keys\"]\n",
    "categorical_keys = config[\"categorical_keys\"] if config[\"categorical_keys\"] else []\n",
    "continuous_keys = config[\"continuous_keys\"] if config[\"continuous_keys\"] else []\n",
    "annotations_loc = config[\"annotations_loc\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'io': {'path_mdata': '/cellar/users/aklie/opt/gene_program_evaluation/examples/inference/iPSC_EC/factor_analysis/factor_analysis_100.h5mu',\n",
       "  'path_out': '/cellar/users/aklie/opt/gene_program_evaluation/examples/evaluation/iPSC_EC/factor_analysis/factor_analysis_100',\n",
       "  'data_key': 'rna',\n",
       "  'prog_key': 'factor_analysis_100'},\n",
       " 'categorical_association': {'categorical_keys': ['sample'],\n",
       "  'pseudobulk_key': None,\n",
       "  'test': 'pearsonr',\n",
       "  'mode': 'one_vs_all',\n",
       "  'n_jobs': -1,\n",
       "  'inplace': False},\n",
       " 'perturbation_association': {'groupby_key': 'sample',\n",
       "  'collapse_targets': True,\n",
       "  'pseudobulk': False,\n",
       "  'reference_targets': ['non-targeting'],\n",
       "  'n_jobs': -1,\n",
       "  'inplace': False},\n",
       " 'gene_set_enrichment': {'prog_nam': None,\n",
       "  'organism': 'human',\n",
       "  'libraries': ['Reactome_2022', 'GO_Biological_Process_2023'],\n",
       "  'method': 'fisher',\n",
       "  'database': 'enrichr',\n",
       "  'n_top': 500,\n",
       "  'low_cutoff': 0.0,\n",
       "  'n_jobs': -1,\n",
       "  'inplace': False,\n",
       "  'user_geneset': None,\n",
       "  'max_size': 500,\n",
       "  'min_size': 5},\n",
       " 'trait_enrichment': {'gwas_data': '/cellar/users/aklie/opt/gene_program_evaluation/smk/resources/OpenTargets_L2G_Filtered.csv.gz',\n",
       "  'prog_nam': None,\n",
       "  'library': 'OT_GWAS',\n",
       "  'n_jobs': -1,\n",
       "  'inplace': False,\n",
       "  'key_column': 'trait_efos',\n",
       "  'gene_column': 'gene_name',\n",
       "  'method': 'fisher',\n",
       "  'n_top': 500,\n",
       "  'low_cutoff': 0.0,\n",
       "  'min_size': 5,\n",
       "  'max_size': 2000,\n",
       "  'metadata': '/cellar/users/aklie/opt/gene_program_evaluation/smk/resources/OpenTargets_L2G_Filtered.csv.gz',\n",
       "  'pval_col': 'adj_pval',\n",
       "  'enrich_geneset_id_col': 'term',\n",
       "  'metadata_geneset_id_col': 'trait_efos',\n",
       "  'color_category_col': 'trait_category',\n",
       "  'program_name_col': 'program_name',\n",
       "  'annotation_cols': ['enrichment',\n",
       "   'trait_reported',\n",
       "   'genes',\n",
       "   'study_id',\n",
       "   'pmid']},\n",
       " 'motif_enrichment': {'groupby_key': 'sample',\n",
       "  'motif_file': '/cellar/users/aklie/opt/gene_program_evaluation/src/tests/test_data/motifs.meme',\n",
       "  'seq_file': '/cellar/users/aklie/data/ref/genomes/hg38/hg38.fa',\n",
       "  'loci_files': ['/cellar/users/aklie/opt/gene_program_evaluation/examples/datasets/iPSC_EC/EnhancerPredictions_D0_2024_09_07.txt',\n",
       "   '/cellar/users/aklie/opt/gene_program_evaluation/examples/datasets/iPSC_EC/EnhancerPredictions_sample_D1_2024_09_07.txt',\n",
       "   '/cellar/users/aklie/opt/gene_program_evaluation/examples/datasets/iPSC_EC/EnhancerPredictions_sample_D2_2024_09_07.txt',\n",
       "   '/cellar/users/aklie/opt/gene_program_evaluation/examples/datasets/iPSC_EC/EnhancerPredictions_sample_D3_2024_09_07.txt'],\n",
       "  'names': ['D0', 'sample_D1', 'sample_D2', 'sample_D3'],\n",
       "  'output_loc': None,\n",
       "  'window': 1000,\n",
       "  'threshold': 0.001,\n",
       "  'eps': 0.001,\n",
       "  'reverse_complement': True,\n",
       "  'sig': 0.05,\n",
       "  'num_genes': None,\n",
       "  'correlation': 'pearsonr',\n",
       "  'n_jobs': -1,\n",
       "  'inplace': False},\n",
       " 'explained_variance': {'layer': 'X', 'n_jobs': 1, 'inplace': False}}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load evaluation config\n",
    "evaluation_config = load_config(path_evaluation_config)\n",
    "evaluation_config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "path_evaluation_outs: ['/cellar/users/aklie/opt/gene_program_evaluation/examples/evaluation/iPSC_EC/factor_analysis/factor_analysis_100']\n",
      "path_evaluation_config: /cellar/users/aklie/opt/gene_program_evaluation/examples/evaluation/iPSC_EC/factor_analysis/factor_analysis_100/evaluation_pipeline.yml\n",
      "path_mdata: /cellar/users/aklie/opt/gene_program_evaluation/examples/evaluation/iPSC_EC/factor_analysis/factor_analysis_100/factor_analysis_100.h5mu\n",
      "path_report_out: /cellar/users/aklie/opt/gene_program_evaluation/examples/report/iPSC_EC/factor_analysis/factor_analysis_100\n",
      "prog_keys: ['factor_analysis_100']\n",
      "data_key: rna\n",
      "categorical_keys: ['sample', 'leiden']\n",
      "continuous_keys: ['n_counts']\n",
      "annotations_loc: annotations.csv\n"
     ]
    }
   ],
   "source": [
    "# Log configuration\n",
    "for key, value in config.items():\n",
    "    print((f\"{key}: {value}\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load mudata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/anndata/_core/anndata.py:1899: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/cellar/users/aklie/opt/miniconda3/envs/test_celloracle/lib/python3.10/site-packages/mudata/_core/mudata.py:479: UserWarning: var_names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre>MuData object with n_obs × n_vars = 92284 × 5551\n",
       "  2 modalities\n",
       "    factor_analysis_100:\t92284 x 100\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;species&#x27;, &#x27;gene_count&#x27;, &#x27;tscp_count&#x27;, &#x27;mread_count&#x27;, &#x27;leiden&#x27;, &#x27;n_counts&#x27;\n",
       "      uns:\t&#x27;loadings_genes&#x27;\n",
       "      varm:\t&#x27;loadings&#x27;\n",
       "    rna:\t92284 x 5451\n",
       "      obs:\t&#x27;sample&#x27;, &#x27;species&#x27;, &#x27;gene_count&#x27;, &#x27;tscp_count&#x27;, &#x27;mread_count&#x27;, &#x27;leiden&#x27;, &#x27;n_counts&#x27;\n",
       "      uns:\t&#x27;guide_names&#x27;, &#x27;guide_targets&#x27;\n",
       "      obsm:\t&#x27;X_pca&#x27;, &#x27;X_umap&#x27;, &#x27;guide_assignment&#x27;\n",
       "      layers:\t&#x27;norm10k&#x27;</pre>"
      ],
      "text/plain": [
       "MuData object with n_obs × n_vars = 92284 × 5551\n",
       "  2 modalities\n",
       "    factor_analysis_100:\t92284 x 100\n",
       "      obs:\t'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "      uns:\t'loadings_genes'\n",
       "      varm:\t'loadings'\n",
       "    rna:\t92284 x 5451\n",
       "      obs:\t'sample', 'species', 'gene_count', 'tscp_count', 'mread_count', 'leiden', 'n_counts'\n",
       "      uns:\t'guide_names', 'guide_targets'\n",
       "      obsm:\t'X_pca', 'X_umap', 'guide_assignment'\n",
       "      layers:\t'norm10k'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdata = mudata.read_h5mu(path_mdata)\n",
    "mdata.mod = collections.OrderedDict(sorted(mdata.mod.items()))\n",
    "mdata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `method` and `n_components`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_methods(mdata, data_key=\"rna\"):\n",
    "    methods = {}\n",
    "    n_components = {}\n",
    "    for key in mdata.mod.keys():\n",
    "        method_split = key.split(\"_\")\n",
    "        if len(method_split) > 1:\n",
    "            method = \"_\".join(method_split[:-1])\n",
    "        else:\n",
    "            method = method_split[0]\n",
    "        if method != data_key:\n",
    "            methods[key] = method\n",
    "            n_components[key] = mdata.mod[key].X.shape[1]\n",
    "    return methods, n_components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "res_methods, res_ks = parse_methods(mdata, data_key=data_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'factor_analysis'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_methods[prog_keys[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_ks[prog_keys[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `loadings`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'var_names'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[41], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m res_loadings \u001b[38;5;241m=\u001b[39m \u001b[43mparse_loadings\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdata_key\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/gene_program_evaluation/app/parse.py:29\u001b[0m, in \u001b[0;36mparse_loadings\u001b[0;34m(mdata, data_key)\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m mdata\u001b[38;5;241m.\u001b[39mmod\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[1;32m     25\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;241m!=\u001b[39m data_key:\n\u001b[1;32m     26\u001b[0m         loadings[key] \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(\n\u001b[1;32m     27\u001b[0m             data\u001b[38;5;241m=\u001b[39mmdata\u001b[38;5;241m.\u001b[39mmod[key]\u001b[38;5;241m.\u001b[39mvarm[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mloadings\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[1;32m     28\u001b[0m             index\u001b[38;5;241m=\u001b[39mmdata\u001b[38;5;241m.\u001b[39mmod[key]\u001b[38;5;241m.\u001b[39mvar_names,\n\u001b[0;32m---> 29\u001b[0m             columns\u001b[38;5;241m=\u001b[39m\u001b[43mmdata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmod\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43muns\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mvar_names\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m     30\u001b[0m         )\n\u001b[1;32m     31\u001b[0m         loadings[key]\u001b[38;5;241m.\u001b[39mindex\u001b[38;5;241m.\u001b[39mname \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprogram_name\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     32\u001b[0m         loadings[key]\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mname \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgene_name\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'var_names'"
     ]
    }
   ],
   "source": [
    "res_loadings = parse_loadings(mdata, data_key=data_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'res_loadings' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[42], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mres_loadings\u001b[49m[prog_keys[\u001b[38;5;241m0\u001b[39m]]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'res_loadings' is not defined"
     ]
    }
   ],
   "source": [
    "res_loadings[prog_keys[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `obs_membership`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "res_obs_membership = parse_obs_memberships(mdata, data_key=data_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>program_name</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>51</th>\n",
       "      <th>52</th>\n",
       "      <th>53</th>\n",
       "      <th>54</th>\n",
       "      <th>55</th>\n",
       "      <th>56</th>\n",
       "      <th>57</th>\n",
       "      <th>58</th>\n",
       "      <th>59</th>\n",
       "      <th>60</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>obs_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT-scRNAseq_2kG_11AMDox_1',)</th>\n",
       "      <td>121.749060</td>\n",
       "      <td>64.133135</td>\n",
       "      <td>8.107442</td>\n",
       "      <td>301.635912</td>\n",
       "      <td>11.174598</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>97.718911</td>\n",
       "      <td>336.144644</td>\n",
       "      <td>1.291447</td>\n",
       "      <td>...</td>\n",
       "      <td>233.636156</td>\n",
       "      <td>447.831548</td>\n",
       "      <td>139.534562</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.765844</td>\n",
       "      <td>143.199367</td>\n",
       "      <td>290.182025</td>\n",
       "      <td>34.604045</td>\n",
       "      <td>628.739760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-scRNAseq_2kG_11AMDox_1',)</th>\n",
       "      <td>116.857174</td>\n",
       "      <td>58.951640</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>24.340045</td>\n",
       "      <td>0.291422</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.313101</td>\n",
       "      <td>...</td>\n",
       "      <td>67.100028</td>\n",
       "      <td>415.237080</td>\n",
       "      <td>106.003152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>124.525017</td>\n",
       "      <td>175.423327</td>\n",
       "      <td>289.857663</td>\n",
       "      <td>156.831588</td>\n",
       "      <td>221.619807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-scRNAseq_2kG_11AMDox_1',)</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.599494</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>223.030173</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>136.979013</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>532.648754</td>\n",
       "      <td>204.271224</td>\n",
       "      <td>256.087966</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>225.720921</td>\n",
       "      <td>315.902555</td>\n",
       "      <td>221.630012</td>\n",
       "      <td>191.158832</td>\n",
       "      <td>1508.063771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT-scRNAseq_2kG_11AMDox_1',)</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>122.133827</td>\n",
       "      <td>10.861754</td>\n",
       "      <td>105.185642</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>94.525288</td>\n",
       "      <td>9.906529</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>199.618592</td>\n",
       "      <td>348.723283</td>\n",
       "      <td>118.126613</td>\n",
       "      <td>75.682063</td>\n",
       "      <td>23.327307</td>\n",
       "      <td>46.885769</td>\n",
       "      <td>254.726283</td>\n",
       "      <td>231.582575</td>\n",
       "      <td>376.122800</td>\n",
       "      <td>932.057594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACATTAAGCC-scRNAseq_2kG_11AMDox_1',)</th>\n",
       "      <td>541.275596</td>\n",
       "      <td>37.484053</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>754.398716</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.170416</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>95.477637</td>\n",
       "      <td>264.667966</td>\n",
       "      <td>16.275740</td>\n",
       "      <td>...</td>\n",
       "      <td>391.251173</td>\n",
       "      <td>344.427816</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>346.156259</td>\n",
       "      <td>145.217304</td>\n",
       "      <td>195.982947</td>\n",
       "      <td>79.628793</td>\n",
       "      <td>265.616580</td>\n",
       "      <td>204.519899</td>\n",
       "      <td>901.975519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGCGC-scRNAseq_2kG_2PMDox2_7',)</th>\n",
       "      <td>18.207947</td>\n",
       "      <td>77.236697</td>\n",
       "      <td>37.510989</td>\n",
       "      <td>179.563481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>264.594245</td>\n",
       "      <td>55.222566</td>\n",
       "      <td>58.927143</td>\n",
       "      <td>3.485854</td>\n",
       "      <td>...</td>\n",
       "      <td>196.284818</td>\n",
       "      <td>383.096991</td>\n",
       "      <td>97.975362</td>\n",
       "      <td>47.863879</td>\n",
       "      <td>141.223926</td>\n",
       "      <td>78.169581</td>\n",
       "      <td>48.317410</td>\n",
       "      <td>65.962411</td>\n",
       "      <td>54.581000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('Enhancer-BP-etc-chr7-150704624-150705124:GCAAGTGGGCGCACAAAGGGA:TTTGTTGGTATCACGT-scRNAseq_2kG_2PMDox2_7',)</th>\n",
       "      <td>176.165401</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.781613</td>\n",
       "      <td>88.219071</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>522.186737</td>\n",
       "      <td>337.551912</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.781014</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>210.987869</td>\n",
       "      <td>27.231758</td>\n",
       "      <td>15.451982</td>\n",
       "      <td>220.460153</td>\n",
       "      <td>28.465382</td>\n",
       "      <td>374.087279</td>\n",
       "      <td>277.836118</td>\n",
       "      <td>217.102964</td>\n",
       "      <td>166.132388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC-scRNAseq_2kG_2PMDox2_7',)</th>\n",
       "      <td>128.595762</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>33.366939</td>\n",
       "      <td>299.690733</td>\n",
       "      <td>22.669654</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>904.535637</td>\n",
       "      <td>92.267307</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>73.585602</td>\n",
       "      <td>551.063049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>79.799175</td>\n",
       "      <td>364.806463</td>\n",
       "      <td>42.103565</td>\n",
       "      <td>327.637684</td>\n",
       "      <td>343.052591</td>\n",
       "      <td>409.746455</td>\n",
       "      <td>260.209294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-scRNAseq_2kG_2PMDox2_7',)</th>\n",
       "      <td>47.865646</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.960127</td>\n",
       "      <td>29.554090</td>\n",
       "      <td>16.559151</td>\n",
       "      <td>1.112656</td>\n",
       "      <td>45.628822</td>\n",
       "      <td>17.997079</td>\n",
       "      <td>39.754875</td>\n",
       "      <td>5.333006</td>\n",
       "      <td>...</td>\n",
       "      <td>190.243596</td>\n",
       "      <td>70.161919</td>\n",
       "      <td>19.528940</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>115.283159</td>\n",
       "      <td>277.144962</td>\n",
       "      <td>166.382661</td>\n",
       "      <td>75.062468</td>\n",
       "      <td>217.111685</td>\n",
       "      <td>31.075039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-scRNAseq_2kG_2PMDox2_7',)</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.044296</td>\n",
       "      <td>353.528466</td>\n",
       "      <td>146.943586</td>\n",
       "      <td>41.367928</td>\n",
       "      <td>7.377000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>171.071045</td>\n",
       "      <td>134.442154</td>\n",
       "      <td>48.715768</td>\n",
       "      <td>110.223886</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>114.359513</td>\n",
       "      <td>245.580548</td>\n",
       "      <td>40.379763</td>\n",
       "      <td>36.095628</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214449 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "program_name                                                 1           2  \\\n",
       "obs_name                                                                     \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...  121.749060   64.133135   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...  116.857174   58.951640   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...    0.000000    0.000000   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...    0.000000  122.133827   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...  541.275596   37.484053   \n",
       "...                                                        ...         ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...   18.207947   77.236697   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...  176.165401    0.000000   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...  128.595762    0.000000   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...   47.865646    0.000000   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...    0.000000    0.000000   \n",
       "\n",
       "program_name                                                3           4  \\\n",
       "obs_name                                                                    \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...   8.107442  301.635912   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...   0.000000   24.340045   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...   9.599494    0.000000   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...  10.861754  105.185642   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...   0.000000  754.398716   \n",
       "...                                                       ...         ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...  37.510989  179.563481   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...   3.781613   88.219071   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...  33.366939  299.690733   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...  12.960127   29.554090   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...   0.000000    0.000000   \n",
       "\n",
       "program_name                                                5         6  \\\n",
       "obs_name                                                                  \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...  11.174598  0.000000   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...   0.291422  0.000000   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...   0.000000  0.000000   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...   0.000000  0.000000   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...   0.000000  1.170416   \n",
       "...                                                       ...       ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...   0.000000  0.000000   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...   0.000000  0.000000   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...  22.669654  0.000000   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...  16.559151  1.112656   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...   0.000000  1.044296   \n",
       "\n",
       "program_name                                                 7           8  \\\n",
       "obs_name                                                                     \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...    0.000000   97.718911   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...    0.000000    0.000000   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...  223.030173    0.000000   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...    0.000000   94.525288   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...    0.000000   95.477637   \n",
       "...                                                        ...         ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...  264.594245   55.222566   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...  522.186737  337.551912   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...  904.535637   92.267307   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...   45.628822   17.997079   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...  353.528466  146.943586   \n",
       "\n",
       "program_name                                                 9         10  \\\n",
       "obs_name                                                                    \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...  336.144644   1.291447   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...    0.000000   3.313101   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...  136.979013   0.000000   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...    9.906529   0.000000   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...  264.667966  16.275740   \n",
       "...                                                        ...        ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...   58.927143   3.485854   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...    0.000000   5.781014   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...    0.000000   0.000000   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...   39.754875   5.333006   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...   41.367928   7.377000   \n",
       "\n",
       "program_name                                        ...          51  \\\n",
       "obs_name                                            ...               \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...  ...  233.636156   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...  ...   67.100028   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...  ...  532.648754   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...  ...  199.618592   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...  ...  391.251173   \n",
       "...                                                 ...         ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...  ...  196.284818   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...  ...    0.000000   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...  ...   73.585602   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...  ...  190.243596   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...  ...    0.000000   \n",
       "\n",
       "program_name                                                52          53  \\\n",
       "obs_name                                                                     \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...  447.831548  139.534562   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...  415.237080  106.003152   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...  204.271224  256.087966   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...  348.723283  118.126613   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...  344.427816    0.000000   \n",
       "...                                                        ...         ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...  383.096991   97.975362   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...  210.987869   27.231758   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...  551.063049    0.000000   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...   70.161919   19.528940   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...  171.071045  134.442154   \n",
       "\n",
       "program_name                                                54          55  \\\n",
       "obs_name                                                                     \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...    0.000000    0.000000   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...    0.000000    0.000000   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...    0.000000    0.000000   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...   75.682063   23.327307   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...  346.156259  145.217304   \n",
       "...                                                        ...         ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...   47.863879  141.223926   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...   15.451982  220.460153   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...   79.799175  364.806463   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...    0.000000  115.283159   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...   48.715768  110.223886   \n",
       "\n",
       "program_name                                                56          57  \\\n",
       "obs_name                                                                     \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...   20.765844  143.199367   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...  124.525017  175.423327   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...  225.720921  315.902555   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...   46.885769  254.726283   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...  195.982947   79.628793   \n",
       "...                                                        ...         ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...   78.169581   48.317410   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...   28.465382  374.087279   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...   42.103565  327.637684   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...  277.144962  166.382661   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...    0.000000  114.359513   \n",
       "\n",
       "program_name                                                58          59  \\\n",
       "obs_name                                                                     \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...  290.182025   34.604045   \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...  289.857663  156.831588   \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...  221.630012  191.158832   \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...  231.582575  376.122800   \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...  265.616580  204.519899   \n",
       "...                                                        ...         ...   \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...   65.962411   54.581000   \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...  277.836118  217.102964   \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...  343.052591  409.746455   \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...   75.062468  217.111685   \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...  245.580548   40.379763   \n",
       "\n",
       "program_name                                                 60  \n",
       "obs_name                                                         \n",
       "('ANKEF1:GAAGGGACATCATTCACGCCT:AAACCCAAGAAGTCAT...   628.739760  \n",
       "('MTRR:GTGGTCCTGGGTACCGAGCAT:AAACCCAAGAGGACTC-s...   221.619807  \n",
       "('JAG1:GATGCGCCCTGCCCGGCGTGC:AAACCCACAATCGCAT-s...  1508.063771  \n",
       "('GOLPH3L:GGAAGTTTGTGCTCTCTGCG:AAACCCACACCAGCGT...   932.057594  \n",
       "('ARHGEF15-TSS2:GACCTACTGCAGAGTTAGGG:AAACCCACAT...   901.975519  \n",
       "...                                                         ...  \n",
       "('ARHGAP26:GCTCGGAAGTGCGGACTATCG:TTTGTTGCAGTTGC...     0.000000  \n",
       "('Enhancer-BP-etc-chr7-150704624-150705124:GCAA...   166.132388  \n",
       "('ERGIC3:GCATGGGGACCGGCCGGAAAG:TTTGTTGGTTCAGCGC...   260.209294  \n",
       "('MMRN2:GTCTTCAAGTTAAATCTCAGG:TTTGTTGTCTATCACT-...    31.075039  \n",
       "('AGTRAP:GCCACCTTCAGGTTCACAGC:TTTGTTGTCTGGAGAG-...    36.095628  \n",
       "\n",
       "[214449 rows x 60 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_obs_membership[prog_keys[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `categorical_associations`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_categorical_associations(dirs, prog_keys):\n",
    "    categorical_associations_results = {}\n",
    "    categorical_associations_posthoc = {}\n",
    "    for dir, prog_key in zip(dirs, prog_keys):\n",
    "        \n",
    "        # Initialize dictionaries \n",
    "        categorical_associations_results[prog_key] = {}\n",
    "        categorical_associations_posthoc[prog_key] = {}\n",
    "\n",
    "        # Load association results\n",
    "        categorical_association_files = glob.glob(os.path.join(dir, f\"{prog_key}_*_association_results.txt\"))\n",
    "        for categorical_association_file in categorical_association_files:\n",
    "            categorical_association_df = pd.read_csv(categorical_association_file, sep=\"\\t\")\n",
    "            categorical_key = categorical_association_file.split(f\"{prog_key}_\")[1].split(\"_association_results.txt\")[0]\n",
    "            categorical_associations_results[prog_key][categorical_key] = categorical_association_df\n",
    "        \n",
    "        # Load association posthoc results\n",
    "        categorical_association_posthoc_files = glob.glob(os.path.join(dir, f\"{prog_key}_*_association_posthoc.txt\"))\n",
    "        for categorical_association_posthoc_file in categorical_association_posthoc_files:\n",
    "            categorical_association_posthoc_df = pd.read_csv(categorical_association_posthoc_file, sep=\"\\t\")\n",
    "            categorical_key = categorical_association_posthoc_file.split(f\"{prog_key}_\")[1].split(\"_association_posthoc.txt\")[0]\n",
    "            categorical_associations_posthoc[prog_key][categorical_key] = categorical_association_posthoc_df\n",
    "            \n",
    "    return categorical_associations_results, categorical_associations_posthoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "res_categorical_associations_results, res_categorical_associations_posthoc = parse_categorical_associations(path_evaluation_outs, prog_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Topyfic_5'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "default_run = list(res_categorical_associations_results.keys())[0]\n",
    "default_run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sample':     sample_kruskall_wallis_stat  sample_kruskall_wallis_pval  \\\n",
       " 0                  66337.004472                          0.0   \n",
       " 1                  49168.022222                          0.0   \n",
       " 2                  47862.731952                          0.0   \n",
       " 3                   9378.180853                          0.0   \n",
       " 4                  19112.311624                          0.0   \n",
       " 5                  51065.037118                          0.0   \n",
       " 6                  42833.176281                          0.0   \n",
       " 7                  56226.908180                          0.0   \n",
       " 8                   3775.306938                          0.0   \n",
       " 9                  12637.036106                          0.0   \n",
       " 10                 32736.316857                          0.0   \n",
       " 11                 28291.546963                          0.0   \n",
       " \n",
       "           program_name  \n",
       " 0    iPSC_EC_5_Topic_1  \n",
       " 1    iPSC_EC_5_Topic_2  \n",
       " 2    iPSC_EC_5_Topic_3  \n",
       " 3    iPSC_EC_5_Topic_4  \n",
       " 4    iPSC_EC_5_Topic_5  \n",
       " 5    iPSC_EC_5_Topic_6  \n",
       " 6    iPSC_EC_5_Topic_7  \n",
       " 7    iPSC_EC_5_Topic_8  \n",
       " 8    iPSC_EC_5_Topic_9  \n",
       " 9   iPSC_EC_5_Topic_10  \n",
       " 10  iPSC_EC_5_Topic_11  \n",
       " 11  iPSC_EC_5_Topic_12  }"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_categorical_associations_results[prog_keys[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Topyfic_5'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[25], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mres_categorical_associations_results\u001b[49m\u001b[43m[\u001b[49m\u001b[43mprog_keys\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[43mdefault_run\u001b[49m\u001b[43m]\u001b[49m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'Topyfic_5'"
     ]
    }
   ],
   "source": [
    "res_categorical_associations_results[prog_keys[0]][default_run]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'batch'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mres_categorical_associations_results\u001b[49m\u001b[43m[\u001b[49m\u001b[43mprog_keys\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mbatch\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'batch'"
     ]
    }
   ],
   "source": [
    "res_categorical_associations_results[prog_keys[0]][\"batch\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `perturbation_association_results`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_perturbation_associations(\n",
    "    dirs, \n",
    "    prog_keys,\n",
    "    stratification_key=None\n",
    "):\n",
    "    perturbation_associations = {}\n",
    "    for dir, prog_key in zip(dirs, prog_keys):\n",
    "        if prog_key not in perturbation_associations:\n",
    "            perturbation_associations[prog_key] = {}\n",
    "        perturbation_associations[prog_key][\"results\"] = {}\n",
    "        perturbation_associations[prog_key][\"gene_guides\"] = []\n",
    "        perturbation_associations[prog_key][\"stratification_keys\"] = []\n",
    "        perturbation_associations[prog_key][\"level_keys\"] = []\n",
    "        perturbation_association_files = glob.glob(os.path.join(dir, f\"{prog_key}_*_perturbation_association.txt\"))\n",
    "        for perturbation_association_file in perturbation_association_files:\n",
    "            gene_guide = perturbation_association_file.split(f\"{prog_key}_\")[1].split(\"_\")[0]\n",
    "            if not stratification_key:\n",
    "                curr_stratification_key = \"global\"\n",
    "                curr_level_key = \"global\"\n",
    "            else:\n",
    "                curr_stratification_key = stratification_key\n",
    "                curr_level_key = perturbation_association_file.split(f\"{prog_key}_{gene_guide}_{curr_stratification_key}_\")[1].split(\"_perturbation_association.txt\")[0]\n",
    "            print(f\"Gene/guide: {gene_guide}, Stratification key: {curr_stratification_key}, Level key: {curr_level_key}\")\n",
    "            df = pd.read_csv(perturbation_association_file, sep=\"\\t\")\n",
    "            perturbation_associations[prog_key][\"results\"][f\"{gene_guide}_{curr_stratification_key}_{curr_level_key}\"] = df\n",
    "            perturbation_associations[prog_key][\"gene_guides\"].append(gene_guide)\n",
    "            perturbation_associations[prog_key][\"stratification_keys\"].append(curr_stratification_key)\n",
    "            perturbation_associations[prog_key][\"level_keys\"].append(curr_level_key)\n",
    "    return perturbation_associations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene/guide: gene, Stratification key: global, Level key: global\n"
     ]
    }
   ],
   "source": [
    "pertubation_associations = parse_perturbation_associations(\n",
    "    path_evaluation_outs, \n",
    "    prog_keys, \n",
    "    stratification_key=evaluation_config[\"perturbation_association\"][\"groupby_key\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_name</th>\n",
       "      <th>program_name</th>\n",
       "      <th>stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCC10</td>\n",
       "      <td>5</td>\n",
       "      <td>0.57549</td>\n",
       "      <td>1.510200e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCG4</td>\n",
       "      <td>56</td>\n",
       "      <td>0.23005</td>\n",
       "      <td>5.365200e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABHD8</td>\n",
       "      <td>22</td>\n",
       "      <td>0.45855</td>\n",
       "      <td>2.907600e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACAD11</td>\n",
       "      <td>15</td>\n",
       "      <td>0.64903</td>\n",
       "      <td>6.916700e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACAD11</td>\n",
       "      <td>30</td>\n",
       "      <td>0.59919</td>\n",
       "      <td>4.800500e-12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  target_name  program_name     stat          pval\n",
       "0      ABCC10             5  0.57549  1.510200e-04\n",
       "1       ABCG4            56  0.23005  5.365200e-05\n",
       "2       ABHD8            22  0.45855  2.907600e-07\n",
       "3      ACAD11            15  0.64903  6.916700e-08\n",
       "4      ACAD11            30  0.59919  4.800500e-12"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_gene_guide = pertubation_associations[prog_keys[0]][\"gene_guide\"][0]\n",
    "first_stratification_key = pertubation_associations[prog_keys[0]][\"stratification_key\"][0]\n",
    "first_level_key = pertubation_associations[prog_keys[0]][\"level_key\"][0]\n",
    "res_pertubation_association = pertubation_associations[prog_keys[0]][\"results\"][f\"{first_gene_guide}_{first_stratification_key}_{first_level_key}\"]\n",
    "res_pertubation_association.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['gene_global_global'])"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pertubation_associations[prog_keys[0]][\"results\"].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['results', 'gene_guide', 'stratification_key', 'level_key'])"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pertubation_associations[prog_keys[0]].keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `explained_variance`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_explained_variance(dirs, prog_keys):\n",
    "    explained_variance_ratios = {}\n",
    "    for dir, prog_key in zip(dirs, prog_keys):\n",
    "        try:\n",
    "            explained_variance_ratio_file = os.path.join(dir, f\"{prog_key}_variance_explained_ratio.txt\")\n",
    "            df = pd.read_csv(explained_variance_ratio_file, sep=\"\\t\")\n",
    "            explained_variance_ratios[prog_key] = df\n",
    "        except FileNotFoundError:\n",
    "            print(f\"File not found: {explained_variance_ratio_file}\")\n",
    "    return explained_variance_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "res_explained_variance = parse_explained_variance(path_evaluation_outs, prog_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program_name</th>\n",
       "      <th>variance_explained_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.018046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>0.026766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0.013374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>0.029387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>0.002182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>0.001648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>0.023513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>0.014861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>0.029700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>0.001846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>0.003365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>0.012964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>0.008912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>0.000281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>0.015150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>0.003249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>0.003378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>0.004095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>0.007991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>0.005638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>0.003457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>0.022052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>0.014656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>0.000963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>0.022013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>0.019365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>27</td>\n",
       "      <td>0.013334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>28</td>\n",
       "      <td>0.017926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>29</td>\n",
       "      <td>0.013986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>30</td>\n",
       "      <td>0.019294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>31</td>\n",
       "      <td>0.013366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>32</td>\n",
       "      <td>0.004865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>33</td>\n",
       "      <td>0.012393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>34</td>\n",
       "      <td>0.019421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>35</td>\n",
       "      <td>0.018413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>36</td>\n",
       "      <td>0.025662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>37</td>\n",
       "      <td>0.015873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>38</td>\n",
       "      <td>0.024358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>39</td>\n",
       "      <td>0.020362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>40</td>\n",
       "      <td>0.015040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>41</td>\n",
       "      <td>0.023960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>42</td>\n",
       "      <td>0.015509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>43</td>\n",
       "      <td>0.007400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>44</td>\n",
       "      <td>0.012862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>45</td>\n",
       "      <td>0.019790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>46</td>\n",
       "      <td>0.023772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>47</td>\n",
       "      <td>0.014547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>48</td>\n",
       "      <td>0.023801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>49</td>\n",
       "      <td>0.020275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>50</td>\n",
       "      <td>0.019748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>51</td>\n",
       "      <td>0.023902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>52</td>\n",
       "      <td>0.020745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>53</td>\n",
       "      <td>0.015504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>54</td>\n",
       "      <td>0.025501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>55</td>\n",
       "      <td>0.010647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>56</td>\n",
       "      <td>0.022776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>57</td>\n",
       "      <td>0.022585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>58</td>\n",
       "      <td>0.013054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>59</td>\n",
       "      <td>0.007343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>60</td>\n",
       "      <td>0.030428</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    program_name  variance_explained_ratio\n",
       "0              1                  0.018046\n",
       "1              2                  0.026766\n",
       "2              3                  0.013374\n",
       "3              4                  0.029387\n",
       "4              5                  0.002182\n",
       "5              6                  0.001648\n",
       "6              7                  0.023513\n",
       "7              8                  0.014861\n",
       "8              9                  0.029700\n",
       "9             10                  0.001846\n",
       "10            11                  0.003365\n",
       "11            12                  0.012964\n",
       "12            13                  0.008912\n",
       "13            14                  0.000281\n",
       "14            15                  0.015150\n",
       "15            16                  0.003249\n",
       "16            17                  0.003378\n",
       "17            18                  0.004095\n",
       "18            19                  0.007991\n",
       "19            20                  0.005638\n",
       "20            21                  0.003457\n",
       "21            22                  0.022052\n",
       "22            23                  0.014656\n",
       "23            24                  0.000963\n",
       "24            25                  0.022013\n",
       "25            26                  0.019365\n",
       "26            27                  0.013334\n",
       "27            28                  0.017926\n",
       "28            29                  0.013986\n",
       "29            30                  0.019294\n",
       "30            31                  0.013366\n",
       "31            32                  0.004865\n",
       "32            33                  0.012393\n",
       "33            34                  0.019421\n",
       "34            35                  0.018413\n",
       "35            36                  0.025662\n",
       "36            37                  0.015873\n",
       "37            38                  0.024358\n",
       "38            39                  0.020362\n",
       "39            40                  0.015040\n",
       "40            41                  0.023960\n",
       "41            42                  0.015509\n",
       "42            43                  0.007400\n",
       "43            44                  0.012862\n",
       "44            45                  0.019790\n",
       "45            46                  0.023772\n",
       "46            47                  0.014547\n",
       "47            48                  0.023801\n",
       "48            49                  0.020275\n",
       "49            50                  0.019748\n",
       "50            51                  0.023902\n",
       "51            52                  0.020745\n",
       "52            53                  0.015504\n",
       "53            54                  0.025501\n",
       "54            55                  0.010647\n",
       "55            56                  0.022776\n",
       "56            57                  0.022585\n",
       "57            58                  0.013054\n",
       "58            59                  0.007343\n",
       "59            60                  0.030428"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_explained_variance[prog_keys[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `geneset_enrichments`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_geneset_enrichments(dirs, prog_keys):\n",
    "    geneset_enrichments = {}\n",
    "    for dir, prog_key in zip(dirs, prog_keys):\n",
    "        if prog_key not in geneset_enrichments:\n",
    "            geneset_enrichments[prog_key] = {}\n",
    "        geneset_enrichments[prog_key][\"results\"] = {}\n",
    "        geneset_enrichments[prog_key][\"libraries\"] = []\n",
    "        geneset_enrichments[prog_key][\"methods\"] = []\n",
    "        geneset_enrichment_files = glob.glob(os.path.join(dir, f\"{prog_key}_*_geneset_enrichment.txt\"))\n",
    "        for gene_set_enrichment_file in geneset_enrichment_files:\n",
    "            method = gene_set_enrichment_file.split(\"_geneset_enrichment.txt\")[0].split(\"_\")[-1]\n",
    "            library = gene_set_enrichment_file.split(f\"{prog_key}_\")[1].split(f\"_{method}_geneset_enrichment.txt\")[0]\n",
    "            print(f\"Library: {library}, Method: {method}\")\n",
    "            df = pd.read_csv(gene_set_enrichment_file, sep=\"\\t\")\n",
    "            geneset_enrichments[prog_key][\"results\"][f\"{library}_{method}\"] = df\n",
    "            geneset_enrichments[prog_key][\"libraries\"].append(library)\n",
    "            geneset_enrichments[prog_key][\"methods\"].append(method)\n",
    "    return geneset_enrichments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Library: Reactome_2022, Method: fisher\n",
      "Library: GO_Biological_Process_2023, Method: fisher\n"
     ]
    }
   ],
   "source": [
    "geneset_enrichments = parse_geneset_enrichments(path_evaluation_outs, prog_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['Reactome_2022_fisher', 'GO_Biological_Process_2023_fisher'])"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneset_enrichments[prog_keys[0]][\"results\"].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GO_Biological_Process_2023', 'Reactome_2022']"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find all unique libraries in list in one line\n",
    "list(set(geneset_enrichments[prog_keys[0]][\"libraries\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "      <th>program_name</th>\n",
       "      <th>overlap_numerator</th>\n",
       "      <th>overlap_denominator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gs_ind_0</td>\n",
       "      <td>2-LTR Circle Formation R-HSA-164843</td>\n",
       "      <td>0.136777</td>\n",
       "      <td>0.441518</td>\n",
       "      <td>9.708661</td>\n",
       "      <td>19.314409</td>\n",
       "      <td>HMGA1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gs_ind_0</td>\n",
       "      <td>ABC Transporter Disorders R-HSA-5619084</td>\n",
       "      <td>0.165154</td>\n",
       "      <td>0.467384</td>\n",
       "      <td>2.494413</td>\n",
       "      <td>4.492128</td>\n",
       "      <td>UBC;PSMA7;PSMD8</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gs_ind_0</td>\n",
       "      <td>ABC-family Proteins Mediated Transport R-HSA-3...</td>\n",
       "      <td>0.038397</td>\n",
       "      <td>0.344295</td>\n",
       "      <td>2.988335</td>\n",
       "      <td>9.741281</td>\n",
       "      <td>PSMD8;PSMA7;EIF2S3;UBC;EIF2S2</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gs_ind_0</td>\n",
       "      <td>ADORA2B Mediated Anti-Inflammatory Cytokine Pr...</td>\n",
       "      <td>0.911332</td>\n",
       "      <td>0.934492</td>\n",
       "      <td>0.620191</td>\n",
       "      <td>0.057583</td>\n",
       "      <td>ADM2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gs_ind_0</td>\n",
       "      <td>ALK Mutants Bind TKIs R-HSA-9700645</td>\n",
       "      <td>0.198016</td>\n",
       "      <td>0.497816</td>\n",
       "      <td>6.599370</td>\n",
       "      <td>10.687067</td>\n",
       "      <td>NPM1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene_set                                               Term   P-value  \\\n",
       "0  gs_ind_0                2-LTR Circle Formation R-HSA-164843  0.136777   \n",
       "1  gs_ind_0            ABC Transporter Disorders R-HSA-5619084  0.165154   \n",
       "2  gs_ind_0  ABC-family Proteins Mediated Transport R-HSA-3...  0.038397   \n",
       "3  gs_ind_0  ADORA2B Mediated Anti-Inflammatory Cytokine Pr...  0.911332   \n",
       "4  gs_ind_0                ALK Mutants Bind TKIs R-HSA-9700645  0.198016   \n",
       "\n",
       "   Adjusted P-value  Odds Ratio  Combined Score  \\\n",
       "0          0.441518    9.708661       19.314409   \n",
       "1          0.467384    2.494413        4.492128   \n",
       "2          0.344295    2.988335        9.741281   \n",
       "3          0.934492    0.620191        0.057583   \n",
       "4          0.497816    6.599370       10.687067   \n",
       "\n",
       "                           Genes  program_name  overlap_numerator  \\\n",
       "0                          HMGA1             1                  1   \n",
       "1                UBC;PSMA7;PSMD8             1                  3   \n",
       "2  PSMD8;PSMA7;EIF2S3;UBC;EIF2S2             1                  5   \n",
       "3                           ADM2             1                  1   \n",
       "4                           NPM1             1                  1   \n",
       "\n",
       "   overlap_denominator  \n",
       "0                    8  \n",
       "1                   77  \n",
       "2                  102  \n",
       "3                  131  \n",
       "4                   12  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_method = geneset_enrichments[prog_keys[0]][\"methods\"][0]\n",
    "first_library = geneset_enrichments[prog_keys[0]][\"libraries\"][0]\n",
    "res_geneset_enrichments = geneset_enrichments[prog_keys[0]][\"results\"][f\"{first_library}_{first_method}\"]\n",
    "res_geneset_enrichments.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `motif_enrichments`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_motif_enrichments(\n",
    "    dirs, \n",
    "    prog_keys,\n",
    "    stratification_key=None\n",
    "):\n",
    "    motif_enrichments = {}\n",
    "    for dir, prog_key in zip(dirs, prog_keys):\n",
    "        if prog_key not in motif_enrichments:\n",
    "            motif_enrichments[prog_key] = {}\n",
    "        motif_enrichments[prog_key][\"results\"] = {}\n",
    "        motif_enrichments[prog_key][\"E_P_types\"] = []\n",
    "        motif_enrichments[prog_key][\"databases\"] = []\n",
    "        motif_enrichments[prog_key][\"test_types\"] = []\n",
    "        motif_enrichments[prog_key][\"stratification_keys\"] = []\n",
    "        motif_enrichments[prog_key][\"level_keys\"] = []\n",
    "        motif_enrichment_files = glob.glob(os.path.join(dir, f\"{prog_key}_*_motif_enrichment.txt\"))\n",
    "        for motif_enrichment_file in motif_enrichment_files:\n",
    "            E_P_type = motif_enrichment_file.split(f\"{prog_key}_\")[1].split(\"_\")[0]\n",
    "            database = motif_enrichment_file.split(f\"{prog_key}_{E_P_type}_\")[1].split(\"_\")[0]\n",
    "            test_type = motif_enrichment_file.split(f\"{prog_key}_{E_P_type}_{database}_\")[1].split(\"_\")[0]\n",
    "            if not stratification_key:\n",
    "                curr_stratification_key = \"global\"\n",
    "                curr_level_key = \"global\"\n",
    "            else:\n",
    "                curr_level_key = motif_enrichment_file.split(f\"{prog_key}_{E_P_type}_{database}_{test_type}_{curr_stratification_key}_\")[1].split(\"_motif_enrichment.txt\")[0]\n",
    "                curr_stratification_key = stratification_key\n",
    "            print(f\"E_P_type: {E_P_type}, Database: {database}, Test type: {test_type}, Stratification key: {curr_stratification_key}, Level key: {curr_level_key}\")\n",
    "            df = pd.read_csv(motif_enrichment_file, sep=\"\\t\")\n",
    "            motif_enrichments[prog_key][\"results\"][f\"{E_P_type}_{database}_{test_type}_{curr_stratification_key}_{curr_level_key}\"] = df\n",
    "            motif_enrichments[prog_key][\"E_P_types\"].append(E_P_type)\n",
    "            motif_enrichments[prog_key][\"databases\"].append(database)\n",
    "            motif_enrichments[prog_key][\"test_types\"].append(test_type)\n",
    "            motif_enrichments[prog_key][\"stratification_keys\"].append(curr_stratification_key)\n",
    "            motif_enrichments[prog_key][\"level_keys\"].append(curr_level_key)\n",
    "    \n",
    "    return motif_enrichments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E_P_type: promoter, Database: hocomoco, Test type: ttest, Stratification key: global, Level key: global\n",
      "E_P_type: enhancer, Database: hocomoco, Test type: ttest, Stratification key: global, Level key: global\n"
     ]
    }
   ],
   "source": [
    "res_motif_enrichment = parse_motif_enrichments(\n",
    "    path_evaluation_outs, \n",
    "    prog_keys,\n",
    "    stratification_key=evaluation_config[\"motif_enrichment\"][\"groupby_key\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>program_name</th>\n",
       "      <th>motif</th>\n",
       "      <th>stat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>AHR</td>\n",
       "      <td>1.594955</td>\n",
       "      <td>0.044631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>AHR</td>\n",
       "      <td>1.242518</td>\n",
       "      <td>0.351685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>AHR</td>\n",
       "      <td>0.901666</td>\n",
       "      <td>0.681555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12</td>\n",
       "      <td>AHR</td>\n",
       "      <td>1.204339</td>\n",
       "      <td>0.446282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>AHR</td>\n",
       "      <td>0.949163</td>\n",
       "      <td>0.829277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40615</th>\n",
       "      <td>6</td>\n",
       "      <td>ZSCA4</td>\n",
       "      <td>0.742726</td>\n",
       "      <td>0.234976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40616</th>\n",
       "      <td>60</td>\n",
       "      <td>ZSCA4</td>\n",
       "      <td>1.113110</td>\n",
       "      <td>0.668214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40617</th>\n",
       "      <td>7</td>\n",
       "      <td>ZSCA4</td>\n",
       "      <td>0.842898</td>\n",
       "      <td>0.488940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40618</th>\n",
       "      <td>8</td>\n",
       "      <td>ZSCA4</td>\n",
       "      <td>0.982363</td>\n",
       "      <td>0.936049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40619</th>\n",
       "      <td>9</td>\n",
       "      <td>ZSCA4</td>\n",
       "      <td>0.460275</td>\n",
       "      <td>0.000282</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40620 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       program_name  motif      stat      pval\n",
       "0                 1    AHR  1.594955  0.044631\n",
       "1                10    AHR  1.242518  0.351685\n",
       "2                11    AHR  0.901666  0.681555\n",
       "3                12    AHR  1.204339  0.446282\n",
       "4                13    AHR  0.949163  0.829277\n",
       "...             ...    ...       ...       ...\n",
       "40615             6  ZSCA4  0.742726  0.234976\n",
       "40616            60  ZSCA4  1.113110  0.668214\n",
       "40617             7  ZSCA4  0.842898  0.488940\n",
       "40618             8  ZSCA4  0.982363  0.936049\n",
       "40619             9  ZSCA4  0.460275  0.000282\n",
       "\n",
       "[40620 rows x 4 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_E_P_type = res_motif_enrichment[prog_keys[0]][\"E_P_types\"][0]\n",
    "first_database = res_motif_enrichment[prog_keys[0]][\"databases\"][0]\n",
    "first_test_type = res_motif_enrichment[prog_keys[0]][\"test_types\"][0]\n",
    "first_stratification_key = res_motif_enrichment[prog_keys[0]][\"stratification_keys\"][0]\n",
    "first_level_key = res_motif_enrichment[prog_keys[0]][\"level_keys\"][0]\n",
    "res_motif_enrichment[prog_keys[0]][\"results\"][f\"{first_E_P_type}_{first_database}_{first_test_type}_{first_stratification_key}_{first_level_key}\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['promoter_hocomoco_ttest_global_global', 'enhancer_hocomoco_ttest_global_global'])"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_motif_enrichment[prog_keys[0]][\"results\"].keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `trait_enrichments`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Same as geneset enrichmetns\n",
    "def parse_trait_enrichments(dirs, prog_keys):\n",
    "    trait_enrichments = {}\n",
    "    for dir, prog_key in zip(dirs, prog_keys):\n",
    "        if prog_key not in trait_enrichments:\n",
    "            trait_enrichments[prog_key] = {}\n",
    "        trait_enrichments[prog_key][\"results\"] = {}\n",
    "        trait_enrichments[prog_key][\"databases\"] = []\n",
    "        trait_enrichments[prog_key][\"methods\"] = []\n",
    "        trait_enrichment_files = glob.glob(os.path.join(dir, f\"{prog_key}_*_trait_enrichment.txt\"))\n",
    "        for trait_enrichment_file in trait_enrichment_files:\n",
    "            method = trait_enrichment_file.split(\"_trait_enrichment.txt\")[0].split(\"_\")[-1]\n",
    "            database = trait_enrichment_file.split(f\"{prog_key}_\")[1].split(f\"_{method}_trait_enrichment.txt\")[0]\n",
    "            print(f\"Database: {database}, Method: {method}\")\n",
    "            df = pd.read_csv(trait_enrichment_file, sep=\"\\t\")\n",
    "            trait_enrichments[prog_key][\"results\"][f\"{database}_{method}\"] = df\n",
    "            trait_enrichments[prog_key][\"databases\"].append(database)\n",
    "            trait_enrichments[prog_key][\"methods\"].append(method)\n",
    "    return trait_enrichments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Database: OT_GWAS, Method: fisher\n"
     ]
    }
   ],
   "source": [
    "trait_enrichments = parse_trait_enrichments(path_evaluation_outs, prog_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term</th>\n",
       "      <th>adj_pval</th>\n",
       "      <th>trait_efos</th>\n",
       "      <th>trait_category</th>\n",
       "      <th>program_name</th>\n",
       "      <th>enrichment</th>\n",
       "      <th>trait_reported</th>\n",
       "      <th>genes</th>\n",
       "      <th>study_id</th>\n",
       "      <th>pmid</th>\n",
       "      <th>-log10(adj_pval)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_0036273</td>\n",
       "      <td>0.020498</td>\n",
       "      <td>GO_0036273</td>\n",
       "      <td>biological process</td>\n",
       "      <td>19</td>\n",
       "      <td>8.017121</td>\n",
       "      <td>Statin medication</td>\n",
       "      <td>IRF2BP2;PNPLA3;LIPG;HMGCR;LDLR;PIR;VEGFA</td>\n",
       "      <td>FINNGEN_R6_RX_STATIN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.688285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO_0036273</td>\n",
       "      <td>0.057469</td>\n",
       "      <td>GO_0036273</td>\n",
       "      <td>biological process</td>\n",
       "      <td>18</td>\n",
       "      <td>4.473323</td>\n",
       "      <td>Statin medication</td>\n",
       "      <td>MAFB;CDKN2B;TRIB1;HHEX;VEGFA</td>\n",
       "      <td>FINNGEN_R6_RX_STATIN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.240570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EFO_0004875</td>\n",
       "      <td>0.086676</td>\n",
       "      <td>EFO_0004875</td>\n",
       "      <td>biological process</td>\n",
       "      <td>36</td>\n",
       "      <td>1.894045</td>\n",
       "      <td>Highest math class taken (MTAG) [MTAG]</td>\n",
       "      <td>TET2;NR2F2;CTTNBP2NL;PHIP;RYBP;ZBTB38;PURA;ZSW...</td>\n",
       "      <td>GCST006568</td>\n",
       "      <td>PMID:30038396</td>\n",
       "      <td>1.062102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO_0006306</td>\n",
       "      <td>0.115787</td>\n",
       "      <td>GO_0006306</td>\n",
       "      <td>biological process</td>\n",
       "      <td>47</td>\n",
       "      <td>2.561764</td>\n",
       "      <td>DNA methylation variation (age effect)</td>\n",
       "      <td>FRY;DACH1;NFIA;TNR;ZNF467;PLCB1;EEA1;SVIL;ATP9A</td>\n",
       "      <td>GCST006660</td>\n",
       "      <td>PMID:30348214</td>\n",
       "      <td>0.936342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EFO_0004337</td>\n",
       "      <td>0.116656</td>\n",
       "      <td>EFO_0004337</td>\n",
       "      <td>biological process</td>\n",
       "      <td>36</td>\n",
       "      <td>3.542221</td>\n",
       "      <td>General cognitive ability</td>\n",
       "      <td>TET2;RBMS1;ATF7IP;FOXO3;NR1D2;JMJD1C</td>\n",
       "      <td>GCST006269</td>\n",
       "      <td>PMID:29844566</td>\n",
       "      <td>0.933092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          term  adj_pval   trait_efos      trait_category  program_name  \\\n",
       "0   GO_0036273  0.020498   GO_0036273  biological process            19   \n",
       "1   GO_0036273  0.057469   GO_0036273  biological process            18   \n",
       "2  EFO_0004875  0.086676  EFO_0004875  biological process            36   \n",
       "3   GO_0006306  0.115787   GO_0006306  biological process            47   \n",
       "4  EFO_0004337  0.116656  EFO_0004337  biological process            36   \n",
       "\n",
       "   enrichment                          trait_reported  \\\n",
       "0    8.017121                       Statin medication   \n",
       "1    4.473323                       Statin medication   \n",
       "2    1.894045  Highest math class taken (MTAG) [MTAG]   \n",
       "3    2.561764  DNA methylation variation (age effect)   \n",
       "4    3.542221               General cognitive ability   \n",
       "\n",
       "                                               genes              study_id  \\\n",
       "0           IRF2BP2;PNPLA3;LIPG;HMGCR;LDLR;PIR;VEGFA  FINNGEN_R6_RX_STATIN   \n",
       "1                       MAFB;CDKN2B;TRIB1;HHEX;VEGFA  FINNGEN_R6_RX_STATIN   \n",
       "2  TET2;NR2F2;CTTNBP2NL;PHIP;RYBP;ZBTB38;PURA;ZSW...            GCST006568   \n",
       "3    FRY;DACH1;NFIA;TNR;ZNF467;PLCB1;EEA1;SVIL;ATP9A            GCST006660   \n",
       "4               TET2;RBMS1;ATF7IP;FOXO3;NR1D2;JMJD1C            GCST006269   \n",
       "\n",
       "            pmid  -log10(adj_pval)  \n",
       "0            NaN          1.688285  \n",
       "1            NaN          1.240570  \n",
       "2  PMID:30038396          1.062102  \n",
       "3  PMID:30348214          0.936342  \n",
       "4  PMID:29844566          0.933092  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_method = trait_enrichments[prog_keys[0]][\"methods\"][0]\n",
    "first_database= trait_enrichments[prog_keys[0]][\"databases\"][0]\n",
    "res_trait_enrichments = trait_enrichments[prog_keys[0]][\"results\"][f\"{first_database}_{first_method}\"]\n",
    "res_trait_enrichments.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term</th>\n",
       "      <th>adj_pval</th>\n",
       "      <th>trait_efos</th>\n",
       "      <th>trait_category</th>\n",
       "      <th>program_name</th>\n",
       "      <th>trait_reported</th>\n",
       "      <th>genes</th>\n",
       "      <th>study_id</th>\n",
       "      <th>pmid</th>\n",
       "      <th>-log10(adj_pval)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [term, adj_pval, trait_efos, trait_category, program_name, trait_reported, genes, study_id, pmid, -log10(adj_pval)]\n",
       "Index: []"
      ]
     },
     "execution_count": 358,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_trait_enrichments.query(f\"program_name == '0'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `software_versions`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "res_software_versions = parse_software_versions(path_evaluation_outs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'evaluation_pipeline_versions': {'gene_program_evaluation': '0.0.1',\n",
       "  'gseapy': '1.1.3',\n",
       "  'mudata': '0.2.3',\n",
       "  'numpy': '1.26.2',\n",
       "  'pandas': '1.5.3',\n",
       "  'scikit-learn': '1.3.2',\n",
       "  'scikit-posthocs': '0.9.0',\n",
       "  'scipy': '1.11.4',\n",
       "  'statsmodels': '0.14.0',\n",
       "  'tangermeme': '0.3.0'}}"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res_software_versions[prog_keys[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `parse`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse(\n",
    "    mdata,\n",
    "    dirs,\n",
    "    data_key=\"rna\",\n",
    "    perturbation_association_stratification_key=None,\n",
    "    motif_enrichment_stratification_key=None\n",
    "):\n",
    "    methods, n_components = parse_methods(mdata, data_key)\n",
    "    loadings = parse_loadings(mdata, data_key)\n",
    "    obs_memberships = parse_obs_memberships(mdata, data_key)\n",
    "    categorical_associations_results, categorical_associations_posthoc = parse_categorical_associations(dirs, methods.keys())\n",
    "    perturbation_associations = parse_perturbation_associations(dirs, methods.keys(), perturbation_association_stratification_key)\n",
    "    geneset_enrichments = parse_geneset_enrichments(dirs, methods.keys())\n",
    "    trait_enrichments = parse_trait_enrichments(dirs, methods.keys())\n",
    "    motif_enrichments = parse_motif_enrichments(dirs, methods.keys(), motif_enrichment_stratification_key)\n",
    "    explained_variance_ratios = parse_explained_variance(dirs, methods.keys())\n",
    "    software_versions = parse_software_versions(dirs)\n",
    "    return {\n",
    "        \"methods\": methods,\n",
    "        \"n_components\": n_components,\n",
    "        \"loadings\": loadings,\n",
    "        \"obs_memberships\": obs_memberships,\n",
    "        \"categorical_associations_results\": categorical_associations_results,\n",
    "        \"categorical_associations_posthoc\": categorical_associations_posthoc,\n",
    "        \"perturbation_associations\": perturbation_associations,\n",
    "        \"geneset_enrichments\": geneset_enrichments,\n",
    "        \"trait_enrichments\": trait_enrichments,\n",
    "        \"motif_enrichments\": motif_enrichments,\n",
    "        \"explained_variance_ratios\": explained_variance_ratios,\n",
    "        \"software_versions\": software_versions,\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'parse_methods' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[35], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m results \u001b[38;5;241m=\u001b[39m \u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmdata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_evaluation_outs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdata_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdata_key\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m    \u001b[49m\u001b[43mperturbation_association_stratification_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mevaluation_config\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mperturbation_association\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgroupby_key\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmotif_enrichment_stratification_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mevaluation_config\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmotif_enrichment\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgroupby_key\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[34], line 8\u001b[0m, in \u001b[0;36mparse\u001b[0;34m(mdata, dirs, data_key, perturbation_association_stratification_key, motif_enrichment_stratification_key)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mparse\u001b[39m(\n\u001b[1;32m      2\u001b[0m     mdata,\n\u001b[1;32m      3\u001b[0m     dirs,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m      6\u001b[0m     motif_enrichment_stratification_key\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m      7\u001b[0m ):\n\u001b[0;32m----> 8\u001b[0m     methods, n_components \u001b[38;5;241m=\u001b[39m \u001b[43mparse_methods\u001b[49m(mdata, data_key)\n\u001b[1;32m      9\u001b[0m     loadings \u001b[38;5;241m=\u001b[39m parse_loadings(mdata, data_key)\n\u001b[1;32m     10\u001b[0m     obs_memberships \u001b[38;5;241m=\u001b[39m parse_obs_memberships(mdata, data_key)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'parse_methods' is not defined"
     ]
    }
   ],
   "source": [
    "results = parse(\n",
    "    mdata,\n",
    "    path_evaluation_outs,\n",
    "    data_key=data_key,\n",
    "    perturbation_association_stratification_key=evaluation_config[\"perturbation_association\"][\"groupby_key\"],\n",
    "    motif_enrichment_stratification_key=evaluation_config[\"motif_enrichment\"][\"groupby_key\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "default_run = \"cNMF\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['methods', 'n_components', 'loadings', 'obs_memberships', 'categorical_associations_results', 'categorical_associations_posthoc', 'perturbation_associations', 'geneset_enrichments', 'trait_enrichments', 'motif_enrichments', 'explained_variance_ratios', 'software_versions'])"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cNMF':     program_name  variance_explained_ratio\n",
       " 0              1                  0.018046\n",
       " 1              2                  0.026766\n",
       " 2              3                  0.013374\n",
       " 3              4                  0.029387\n",
       " 4              5                  0.002182\n",
       " 5              6                  0.001648\n",
       " 6              7                  0.023513\n",
       " 7              8                  0.014861\n",
       " 8              9                  0.029700\n",
       " 9             10                  0.001846\n",
       " 10            11                  0.003365\n",
       " 11            12                  0.012964\n",
       " 12            13                  0.008912\n",
       " 13            14                  0.000281\n",
       " 14            15                  0.015150\n",
       " 15            16                  0.003249\n",
       " 16            17                  0.003378\n",
       " 17            18                  0.004095\n",
       " 18            19                  0.007991\n",
       " 19            20                  0.005638\n",
       " 20            21                  0.003457\n",
       " 21            22                  0.022052\n",
       " 22            23                  0.014656\n",
       " 23            24                  0.000963\n",
       " 24            25                  0.022013\n",
       " 25            26                  0.019365\n",
       " 26            27                  0.013334\n",
       " 27            28                  0.017926\n",
       " 28            29                  0.013986\n",
       " 29            30                  0.019294\n",
       " 30            31                  0.013366\n",
       " 31            32                  0.004865\n",
       " 32            33                  0.012393\n",
       " 33            34                  0.019421\n",
       " 34            35                  0.018413\n",
       " 35            36                  0.025662\n",
       " 36            37                  0.015873\n",
       " 37            38                  0.024358\n",
       " 38            39                  0.020362\n",
       " 39            40                  0.015040\n",
       " 40            41                  0.023960\n",
       " 41            42                  0.015509\n",
       " 42            43                  0.007400\n",
       " 43            44                  0.012862\n",
       " 44            45                  0.019790\n",
       " 45            46                  0.023772\n",
       " 46            47                  0.014547\n",
       " 47            48                  0.023801\n",
       " 48            49                  0.020275\n",
       " 49            50                  0.019748\n",
       " 50            51                  0.023902\n",
       " 51            52                  0.020745\n",
       " 52            53                  0.015504\n",
       " 53            54                  0.025501\n",
       " 54            55                  0.010647\n",
       " 55            56                  0.022776\n",
       " 56            57                  0.022585\n",
       " 57            58                  0.013054\n",
       " 58            59                  0.007343\n",
       " 59            60                  0.030428}"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[\"explained_variance_ratios\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'prog_key' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[88], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m results[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmotif_enrichments\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[43mprog_key\u001b[49m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresults\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mkeys()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'prog_key' is not defined"
     ]
    }
   ],
   "source": [
    "results[\"motif_enrichments\"][prog_key][\"results\"].keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DONE!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10 test_celloracle",
   "language": "python",
   "name": "test_scenicplus-py310"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
